File: src/webapp/client/src/ts/types/databases.ts
import { v4 as uuidv4 } from 'uuid'

export enum DatabaseType {
    Other = 1,
    PostgreSQL = 2,
    MySQL = 3,
    MariaDB = 4,
    SAPHana = 5,
    IBMDb2 = 6,
    Oracle = 7,
    MsSql = 8,
    MsAccess = 9,
    SQLite = 10,
    H2 = 11
}

export function databaseTypeToString(typ : DatabaseType) : string {
    switch (typ) {
        case DatabaseType.PostgreSQL:
            return 'PostgreSQL'
        case DatabaseType.MySQL:
            return 'MySQL'
        case DatabaseType.MariaDB:
            return 'MariaDB'
        case DatabaseType.SAPHana:
            return 'SAP HANA'
        case DatabaseType.IBMDb2:
            return 'IBM DB2'
        case DatabaseType.Oracle:
            return 'Oracle'
        case DatabaseType.MsSql:
            return 'Microsoft SQL'
        case DatabaseType.MsAccess:
            return 'Microsoft Access'
        case DatabaseType.SQLite:
            return 'SQLite'
        case DatabaseType.H2:
            return 'H2'
        default: 
            return 'Unknown'
    }
}

export interface RawDatabase {
    Id                  : number
    EngagementId        : number
    UniqueId            : string
    Name                : string
    Description         : string
    TypeId              : DatabaseType
    OtherType           : string | null
    Version             : string
}

export function createEmptyDatabase() : RawDatabase {
    return {
        Id: -1,
        EngagementId: -1,
        UniqueId: uuidv4(),
        Name: '',
        Description: '',
        TypeId: DatabaseType.Other,
        OtherType: null,
        Version: ''
    }
}

File: src/shared/golang/backend/databases/databases.go
package databases

import (
	"github.com/jmoiron/sqlx"
)

type Database struct {
	Id           int64   `db:"id"`
	EngagementId int64   `db:"engagement_id"`
	UniqueId     string  `db:"unique_id"`
	Name         string  `db:"name"`
	Description  string  `db:"description"`
	TypeId       int32   `db:"type_id"`
	OtherType    *string `db:"other_type"`
	Version      string  `db:"version"`
}

type DatabaseManager struct {
	db *sqlx.DB
}

func CreateDatabaseManager(db *sqlx.DB) *DatabaseManager {
	return &DatabaseManager{
		db: db,
	}
}
